File: 2025/js/day1/part2.js
import { getInput } from "aoc";

async function main() {
//   const input = `L68
// L30
// R48
// L5
// R60
// L55
// L1
// L99
// R14
// L82
// `;
  const input = await getInput(import.meta.dir);

  let dial = 50;
  let passwd = 0;

  for (const line of input.trim().split("\n")) {
    const starts_at_0 = dial === 0;

    const direction = line.substring(0, 1);
    const distance = parseInt(line.substring(1));
    // console.debug(`${direction}, ${distance}`);

    dial = direction === "L" ? dial - distance : dial + distance;
    const cross = Math.floor(dial / 100);
    dial = ((dial % 100) + 100) % 100;
    const ends_at_0 = dial === 0;
    // console.debug(dial);

    if (cross === 0 && ends_at_0) passwd += 1;
    if (cross > 0) passwd += cross;
    if (cross < 0) {
      passwd += Math.abs(cross);
      if (starts_at_0) passwd -= 1;
      if (ends_at_0) passwd += 1;
    }
  }

  console.debug(`passwd: ${passwd}`);
}

const tStart = performance.now();
await main();
const tEnd = performance.now();
console.debug(`Execution time: ${(tEnd - tStart).toFixed(3)} ms.`);

File: 2025/python/day1/part2.py
"""
Solving: https://adventofcode.com/2025/day/1#part2
"""

import re
from typing import Literal

from aoc import get_input, log, time_this


@time_this
def main():
    input = """L68
L30
R48
L5
R60
L55
L1
L99
R14
L82
"""
    input = get_input(__file__)

    dial = 50
    passwd = 0

    for line in input.splitlines():
        start_at_0 = True if dial == 0 else False

        match = re.search(r"^([LR])(\d+)$", line)
        assert match is not None

        direction: Literal["L", "R"] = match.group(1)  # ty:ignore[invalid-assignment]
        distance: int = int(match.group(2))

        dial = dial - distance if direction == "L" else dial + distance
        cross, dial = divmod(dial, 100)

        end_at_0 = True if dial == 0 else False

        if cross == 0 and end_at_0:
            passwd += 1
        if cross > 0:
            passwd += cross
        if cross < 0:
            passwd += abs(cross)
            if start_at_0:
                passwd -= 1
            if end_at_0:
                passwd += 1

        # log.debug(f"{cross=}, {dial=}, {passwd=}")

    log.debug(f"{passwd=}")


if __name__ == "__main__":
    main()
